deps/rabbit/src/rabbit_db_m2k_converter.erl

```diff
@@ -63,65 +124,130 @@ init_copy_to_khepri(StoreId, MigrationId, _Tables, Migrations) ->
+            case ActualMod:delete_from_khepri(Table, Key, State) of
+                {ok, State1} ->
                     {ok, State1};
                 {error, _} = Error ->
                     Error
             end
     end.
 
+-spec finish_copy_to_khepri(State) -> Ret when
+      State :: state(),
+      Ret :: ok.
+%% @private
+
+finish_copy_to_khepri(State) ->
+    {ok, _} = wait_for_all_async_requests(State),
+    ok.
+
+wait_for_all_async_requests(
+  #?MODULE{seq_no = SeqNo,
+           last_acked_seq_no = LastAckedSeqNo} = State) ->
+    case SeqNo - LastAckedSeqNo > 0 of
+        true ->
+            case wait_for_async_requests(State) of
+                {ok, State1} ->
+                    wait_for_all_async_requests(State1);
+                {error, _} = Error ->
+                    Error
+            end;
+        false ->
+            {ok, State}
+    end.
+
+wait_for_async_requests(
+  #?MODULE{store_id = StoreId,
+           seq_no = SeqNo0,
+           async_requests = AsyncRequests0} = State0) ->
+    receive
+        {ra_event, _CurrentLeader, {applied, Correlations}} = AsyncRet ->
+            ok = khepri:handle_async_ret(StoreId, AsyncRet),
+            lists:foldl(
+              fun (_, {error, _} = Error) ->
+                      Error;
+                  ({SeqNo, Resp}, {ok, State}) ->
+                      #?MODULE{async_requests = AsyncRequests,
+                               last_acked_seq_no = LastAcked} = State,
+                      LastAcked1 = erlang:max(SeqNo, LastAcked),
+                      AsyncRequests1 = maps:remove(SeqNo, AsyncRequests),
+                      State1 = State#?MODULE{last_acked_seq_no = LastAcked1,
+                                             async_requests = AsyncRequests1},
+                      case Resp of
+                          ok ->
+                              {ok, State1};
+                          {ok, _} ->
+                              {ok, State1};
+                          {error, _} = Error ->
+                              Error;
+                          {exception, _, _, _} = Exception ->
+                              khepri_machine:handle_tx_exception(
+                                Exception)
+                      end
+              end, {ok, State0}, Correlations);
+        {ra_event,
+         _FromId,
+         {rejected, {not_leader, _MaybeLeader, FailedSeqNo}}} = AsyncRet ->
+            ok = khepri:handle_async_ret(StoreId, AsyncRet),
+            {Fun, AsyncRequests} = maps:take(FailedSeqNo, AsyncRequests0),
+            SeqNo = SeqNo0 + 1,
+            AsyncRequests1 = AsyncRequests#{SeqNo => Fun},
+            State = State0#?MODULE{seq_no = SeqNo,
+                                   async_requests = AsyncRequests1},
+            case Fun(SeqNo) of
+                ok ->
+                    {ok, State};
+                {error, _} = Error ->
+                    Error
+            end
```

Comment:
This code is duplicated, the same exists in `with_correlation_id/2`. Can this be moved to a dedicated function?